### ISSUE TITLE
Add algorithm tuning infrastructure

### ISSUE THREAD

**Issue by kylelutz** [boostorg/compute#357]
Add infrastructure for tuning algorithm parameters to achieve maximum performance on the hardware present at run-time. Many algorithms have variable execution parameters (e.g. `VPT`, `TPB`, etc.) which can be tuned for the specific compute device to yield better performance and allow the algorithms to scale better to different hardware.
##

**Comment by roshanr95** [comment]
What is the status of this now? Has somebody started working on this?

**Comment by kylelutz** [comment]
Yeah, I have an implementation almost ready for this. I'll try to get it pushed sometime this week.

**Comment by kylelutz** [comment]
Initial implementation in PR #434.
